File: tests/service_test.cpp
/*******************************************************************************
 INCLUDES
*******************************************************************************/

#include "cds/storage.h"
#include "cds/service.h"
#include "cds/debug.h"

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cassert>
#include <cstdio>
#include <cmath>
#include <algorithm>
#include <numeric>
#include <thread>
#include <chrono>

/*******************************************************************************
 USINGS
*******************************************************************************/

using std::string;
using std::ofstream;
using std::vector;
using std::min;
using std::to_string;
using std::fabs;
using std::cout;
using std::endl;
using std::move;
using std::min_element;
using std::max_element;
using std::accumulate;
using std::this_thread::sleep_for;
using std::chrono::milliseconds;

/*******************************************************************************
 PARAMETERS
*******************************************************************************/

namespace {
    constexpr bool isVerbose = true;
    constexpr int storageCloseWaitMilli = 5;        // avoid "device is busy" error with rapid mount/unmount calls

    constexpr size_t ramDiskSizeMb = 5000;          // make big enough to contain data
    const string defaultWorkDir = "/tmp/.cds";

    constexpr size_t minFileCount = 1;
    constexpr size_t minRecordCount = 1;
    constexpr size_t minFieldCount = 1;

    constexpr size_t maxFileCount = 10;
    constexpr size_t maxRecordCount = 10;
    constexpr size_t maxFieldCount = 10;
}

/*******************************************************************************
 UTILITY FUNCTIONS
*******************************************************************************/

DataStats generateTestFiles(const string& directory, const int fileCount, const int recordCount, const int fieldCount) {
    if (!checkDirectory(directory)) {
        exit(EXIT_FAILURE);
    }

    if (fileCount < 1 || recordCount < 1 || fieldCount < 1) {
        ERROR("invalid parameters for generateTestFiles()");
        exit(EXIT_FAILURE);
    }

    vector<vector<double>> data(fieldCount);
    int totalRecords = 0;

    for (int i = 1; i <= fileCount; i++) {
        string filename = directory + "/input/test_data_" + to_string(i) + ".csv";
        ofstream file(filename);

        for (int j = 0; j < recordCount; j++) {
            const auto x = static_cast<double>(j % 3) * 3.3;

            for (int k = 0; k < fieldCount; k++) {
                int fieldIndex = k % 3;
                constexpr auto a = 1.1;
                constexpr auto b = 2.2;
                constexpr auto c = 3.3;

                double value = (fieldIndex == 0 ? a : (fieldIndex == 1 ? b : c)) + x;
                file << to_string(value).substr(0, 3);

                if (k < fieldCount - 1) {
                    file << ",";
                }

                data[k].push_back(value);
            }

            file << "\n";
            totalRecords++;
        }

        file.close();
        cout << "Test file created: " << filename << endl;
    }

    DataStats ds(fieldCount);
    ds.recordCount = totalRecords;

    for (int i = 0; i < fieldCount; i++) {
        ds.minimums[i] = *min_element(data[i].begin(), data[i].end());
        ds.maximums[i] = *max_element(data[i].begin(), data[i].end());

        double total = accumulate(data[i].begin(), data[i].end(), 0.0);
        double mean = total / static_cast<double>(data[i].size());
        ds.totals[i] = total;
        ds.means[i] = mean;

        vector<double> sumVec(totalRecords);
        for (int j = 0; j < totalRecords; j++) {
            sumVec[j] = pow(data[i][j] - mean, 2);
        }
        double sum = accumulate(sumVec.begin(), sumVec.end(), 0.0);
        ds.stdDevs[i] = sqrt(sum / static_cast<double>(totalRecords));
    }

    if (totalRecords > 1) {
        for (int i = 0; i < fieldCount; i++) {
            double previous = data[i][0];
            data[i][0] = fabs(data[i][1] - data[i][0]);

            for (int j = 1; j < totalRecords; j++) {
                double current = data[i][j];
                data[i][j] = fabs(current - previous);
                previous = current;
            }
        }

        for (int i = 0; i < fieldCount; i++) {
            ds.deltaMinimums[i] = *min_element(data[i].begin(), data[i].end());
            ds.deltaMaximums[i] = *max_element(data[i].begin(), data[i].end());

            double total = accumulate(data[i].begin(), data[i].end(), 0.0);
            double mean = total / static_cast<double>(data[i].size());
            ds.deltaTotals[i] = total;
            ds.deltaMeans[i] = mean;

            vector<double> sumVec(totalRecords);
            for (int j = 0; j < totalRecords; j++) {
                sumVec[j] = pow(data[i][j] - mean, 2);
            }
            double sum = accumulate(sumVec.begin(), sumVec.end(), 0.0);
            ds.deltaStdDevs[i] = sqrt(sum / static_cast<double>(totalRecords));
        }
    } else {
        for (int i = 0; i < fieldCount; i++) {
            ds.deltaMinimums[i] = data[i][0];
            ds.deltaMaximums[i] = data[i][0];
            ds.deltaMeans[i] = data[i][0];
            ds.deltaTotals[i] = data[i][0];
        }
    }

    return ds;
}

void printDataStats(const DataStats& ds) {
    cout << "DataStats:" << endl;
    cout << "\tField Count: " << ds.fieldCount << endl;
    cout << "\tRecord Count: " << ds.recordCount << endl;

    auto printVector = [](const string& name, const vector<double>& vec) {
        cout << name << ": ";
        for (const auto& val : vec) {
            cout << val << " ";
        }
        cout << endl;
    };

    printVector("\tMinimums", ds.minimums);
    printVector("\tMaximums", ds.maximums);
    printVector("\tTotals", ds.totals);
    printVector("\tMeans", ds.means);
    printVector("\tStandard Deviations", ds.stdDevs);

    printVector("\tDelta Minimums", ds.deltaMinimums);
    printVector("\tDelta Maximums", ds.deltaMaximums);
    printVector("\tDelta Totals", ds.deltaTotals);
    printVector("\tDelta Means", ds.deltaMeans);
    printVector("\tDelta Standard Deviations", ds.deltaStdDevs);

    cout << endl;
}

/*******************************************************************************
 TESTS  // TODO: add more test cases
*******************************************************************************/

void testProcessInputFiles() {
    PRINTLN("\nRunning testProcessInputFiles()...");

    assert(InitStorage(defaultWorkDir.c_str(), ramDiskSizeMb));

    generateTestFiles(defaultWorkDir, 3, 3, 3);

    assert(ProcessInputFiles());

    int recordCount, fieldCount;
    GetFieldAndRecordCount(&recordCount, &fieldCount);
    assert(recordCount == 9);
    assert(fieldCount == 3);

    assert(CloseStorage());
}

void testAnalyzeData(const int fileCount, const int recordCount, const int fieldCount) {
    PRINTLN("\nRunning testAnalyzeData(fileCount: %d, recordCount: %d, fieldCount: %d)...",
        fileCount, recordCount, fieldCount);

    assert(InitStorage(defaultWorkDir.c_str(), ramDiskSizeMb));

    const DataStats expectedStats = generateTestFiles(defaultWorkDir, fileCount, recordCount, fieldCount);
    setWorkingDirectory(defaultWorkDir);

    assert(ProcessInputFiles());
    assert(AnalyzeData());

    int totalRecordCount, actualFieldCount;
    GetFieldAndRecordCount(&totalRecordCount, &actualFieldCount);

    assert(totalRecordCount == fileCount * recordCount);
    assert(actualFieldCount == fieldCount);

    if (const DataStats actualStats = stats::get(); actualStats != expectedStats) {
        printDataStats(actualStats);
        printDataStats(expectedStats);
        sleep_for(milliseconds(20)); // allow time for printing to stdout before assert exits

        assert(actualStats == expectedStats && "actual DataStats does not equal expected DataStats");
    }

    assert(CloseStorage());
}

/*******************************************************************************
 MAIN
*******************************************************************************/

int main() {
    setVerbose(isVerbose);

    testProcessInputFiles();

    for (int i = minFileCount; i <= maxFileCount; i++) {
        for (int j = minRecordCount; j <= maxRecordCount; j++) {
            for (int k = minFieldCount; k <= maxFieldCount; k++) {
                testAnalyzeData(i, j, k);
                sleep_for(milliseconds(storageCloseWaitMilli));
            }
        }
    }

    testAnalyzeData(1, 1000000, 3);
    testAnalyzeData(2, 500000, 3);
    testAnalyzeData(4, 250000, 3);

    printf("All tests completed successfully.\n");
    exit(EXIT_SUCCESS);
}

File: gocds/_test/service_test.go
package _test

import (
	"github.com/stretchr/testify/assert"
	"github.com/tonybillings/gocds/_test"
	"github.com/tonybillings/gocds/cds"
	"testing"
)

func TestProcessInputFiles(t *testing.T) {
	_test.BeginTest(t)
	defer _test.EndTest(t)

	if count := _test.GetInputFileCount(); count != 0 {
		t.Errorf("Expected file count of 0, got %d", count)
		t.FailNow()
	}

	_test.AddTestDataRow()

	if count := _test.GetInputFileCount(); count != 1 {
		t.Errorf("Expected file count of 1, got %d", count)
		t.FailNow()
	}

	if !cds.ProcessInputFiles() {
		t.Error("Failed to process input files")
		t.FailNow()
	}

	if count := _test.GetInputFileCount(); count != 0 {
		t.Errorf("Expected file count of 0, got %d", count)
		t.FailNow()
	}
}

func TestAnalyzeData(t *testing.T) {
	_test.BeginTest(t)
	defer _test.EndTest(t)

	if !cds.AnalyzeData() {
		t.Error("Failed to analyze data with 0 records")
		t.FailNow()
	}

	_test.AddTestDataRow()

	if !cds.ProcessInputFiles() {
		t.Error("Failed to process input files")
		t.FailNow()
	}

	if !cds.AnalyzeData() {
		t.Error("Failed to analyze data with 1 record")
		t.FailNow()
	}
}

func TestGetFieldAndRecordCount(t *testing.T) {
	_test.BeginTest(t)
	defer _test.EndTest(t)

	if fieldCount, recordCount := cds.GetFieldAndRecordCount(); fieldCount != 0 || recordCount != 0 {
		t.Errorf("Unexpected field/record count: expected 0/0, got %d/%d", fieldCount, recordCount)
		t.FailNow()
	}

	_test.AddTestDataRow()

	if !cds.ProcessInputFiles() {
		t.Error("Failed to process input files")
		t.FailNow()
	}

	if !cds.AnalyzeData() {
		t.Error("Failed to analyze data")
		t.FailNow()
	}

	if fieldCount, recordCount := cds.GetFieldAndRecordCount(); fieldCount != 3 || recordCount != 1 {
		t.Errorf("Unexpected field/record count: expected 3/1, got %d/%d", fieldCount, recordCount)
		t.FailNow()
	}

	_test.AddTestDataRow(4)

	if !cds.ProcessInputFiles() {
		t.Error("Failed to process input files")
		t.FailNow()
	}

	if !cds.AnalyzeData() {
		t.Error("Failed to analyze data")
		t.FailNow()
	}

	if fieldCount, recordCount := cds.GetFieldAndRecordCount(); fieldCount != 3 || recordCount != 5 {
		t.Errorf("Unexpected field/record count: expected 3/5, got %d/%d", fieldCount, recordCount)
		t.FailNow()
	}
}

func TestGetStats(t *testing.T) {
	_test.BeginTest(t)
	defer _test.EndTest(t)

	if ds := cds.GetStats(); ds.FieldCount != 0 || ds.RecordCount != 0 {
		t.Errorf("Unexpected field/record count: expected 0/0, got %d/%d", ds.FieldCount, ds.RecordCount)
		t.FailNow()
	}

	_test.AddTestDataRow(333)

	if !cds.ProcessInputFiles() {
		t.Error("Failed to process input files")
		t.FailNow()
	}

	if !cds.AnalyzeData() {
		t.Error("Failed to analyze data")
		t.FailNow()
	}

	if ds := cds.GetStats(); ds.FieldCount != 3 || ds.RecordCount != 333 {
		t.Errorf("Unexpected field/record count: expected 3/333, got %d/%d", ds.FieldCount, ds.RecordCount)
		t.FailNow()
	} else {
		assert.InDelta(t, 1.1, ds.Minimums[0], 0.000001, "unexpected value for Minimums[0]")
		assert.InDelta(t, 2.2, ds.Minimums[1], 0.000001, "unexpected value for Minimums[1]")
		assert.InDelta(t, 3.3, ds.Minimums[2], 0.000001, "unexpected value for Minimums[2]")

		assert.InDelta(t, 7.7, ds.Maximums[0], 0.000001, "unexpected value for Maximums[0]")
		assert.InDelta(t, 8.8, ds.Maximums[1], 0.000001, "unexpected value for Maximums[1]")
		assert.InDelta(t, 9.9, ds.Maximums[2], 0.000001, "unexpected value for Maximums[2]")

		assert.InDelta(t, 1465.2, ds.Totals[0], 0.000001, "unexpected value for Totals[0]")
		assert.InDelta(t, 1831.5, ds.Totals[1], 0.000001, "unexpected value for Totals[1]")
		assert.InDelta(t, 2197.7999999, ds.Totals[2], 0.000001, "unexpected value for Totals[2]")

		assert.InDelta(t, 4.4, ds.Means[0], 0.000001, "unexpected value for Means[0]")
		assert.InDelta(t, 5.5, ds.Means[1], 0.000001, "unexpected value for Means[1]")
		assert.InDelta(t, 6.6, ds.Means[2], 0.000001, "unexpected value for Means[2]")

		assert.InDelta(t, 2.6944387, ds.StdDevs[0], 0.000001, "unexpected value for StdDevs[0]")
		assert.InDelta(t, 2.6944387, ds.StdDevs[1], 0.000001, "unexpected value for StdDevs[1]")
		assert.InDelta(t, 2.6944387, ds.StdDevs[2], 0.000001, "unexpected value for StdDevs[2]")

		assert.InDelta(t, 3.3, ds.DeltaMinimums[0], 0.000001, "unexpected value for DeltaMinimums[0]")
		assert.InDelta(t, 3.3, ds.DeltaMinimums[1], 0.000001, "unexpected value for DeltaMinimums[1]")
		assert.InDelta(t, 3.3, ds.DeltaMinimums[2], 0.000001, "unexpected value for DeltaMinimums[2]")

		assert.InDelta(t, 6.6, ds.DeltaMaximums[0], 0.000001, "unexpected value for DeltaMaximums[0]")
		assert.InDelta(t, 6.6, ds.DeltaMaximums[1], 0.000001, "unexpected value for DeltaMaximums[1]")
		assert.InDelta(t, 6.6, ds.DeltaMaximums[2], 0.000001, "unexpected value for DeltaMaximums[2]")

		assert.InDelta(t, 1461.9, ds.DeltaTotals[0], 0.000001, "unexpected value for DeltaTotals[0]")
		assert.InDelta(t, 1461.9, ds.DeltaTotals[1], 0.000001, "unexpected value for DeltaTotals[1]")
		assert.InDelta(t, 1461.9, ds.DeltaTotals[2], 0.000001, "unexpected value for DeltaTotals[2]")

		assert.InDelta(t, 4.39009, ds.DeltaMeans[0], 0.000001, "unexpected value for DeltaMeans[0]")
		assert.InDelta(t, 4.39009, ds.DeltaMeans[1], 0.000001, "unexpected value for DeltaMeans[1]")
		assert.InDelta(t, 4.39009, ds.DeltaMeans[2], 0.000001, "unexpected value for DeltaMeans[2]")

		assert.InDelta(t, 1.5520956, ds.DeltaStdDevs[0], 0.000001, "unexpected value for DeltaStdDevs[0]")
		assert.InDelta(t, 1.5520956, ds.DeltaStdDevs[1], 0.000001, "unexpected value for DeltaStdDevs[1]")
		assert.InDelta(t, 1.5520956, ds.DeltaStdDevs[2], 0.000001, "unexpected value for DeltaStdDevs[2]")
	}
}
